ENH Sample weights for median_absolute_error

lucyleeow opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Follows from #6217
Addresses last item in #3450

#### What does this implement/fix? Explain your changes.
Add sample_weight to `median_absolute_error`.
Use `sklearn.utils.stats._weighted_percentile` to calculated weighted median as suggested here: https://github.com/scikit-learn/scikit-learn/pull/6217#issuecomment-611930725
Amended `_weighted_percentile` to calculate weighted percentile along axis=0 (for each column) if 2D array input. This is to allow multioutput. Does not change behaviour of `_weighted_percentile` when array 1D. Not sure if this is best way to implement thus will wait for comment before fixing tests.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lucyleeow committed:
weighted percentile 2d

lucyleeow committed:
sample weights none

lucyleeow committed:
impl in median abs err

lucyleeow committed:
lint

lucyleeow committed:
lint

lucyleeow committed:
fix shape

lucyleeow committed:
fix type

lucyleeow committed:
lint

lucyleeow committed:
squeeze

lucyleeow committed:
ndim 0  case

lucyleeow committed:
merge master

lucyleeow committed:
fix text

lucyleeow committed:
take along axis version

lucyleeow committed:
use new fun

lucyleeow committed:
fix format

lucyleeow committed:
spelling

lucyleeow committed:
lint

lucyleeow committed:
lint

lucyleeow committed:
fix logic

lucyleeow committed:
add comment

lucyleeow committed:
add test for multioutput

lucyleeow commented:
Used `np.take_along_axis` to amend `_weighted_percentile` to work for 2D arrays. As `np.take_along_axis` was introduced in numpy v1.15, added the function in `fixes.py` that implements a simplified version of `numpy.take_along_axis` if numpy version < 1.15.

`test_scorer_sample_weight` doesn't work for `neg_median_absolute_error` due to the binary targets (also noted in the original PR https://github.com/scikit-learn/scikit-learn/pull/6217#issuecomment-178378425), so skipped for this scorer. The test works for `neg_median_absolute_error` if targets continuous (e.g., using `make_regression` to generate data). Not sure best way to add/amend to check `neg_median_absolute_error` as well. 

ping @glemaitre :p

NicolasHug commented on sklearn/metrics/_regression.py:
```diff
@@ -387,7 +394,8 @@ def median_absolute_error(y_true, y_pred, *, multioutput='uniform_average'):
     y_type, y_true, y_pred, multioutput = _check_reg_targets(
         y_true, y_pred, multioutput)
-    output_errors = np.median(np.abs(y_pred - y_true), axis=0)
+    output_errors = _weighted_percentile(np.abs(y_pred - y_true),
+                                         sample_weight=sample_weight)
```
I think we should avoid calling `_weighted_percentile` if sample_weight is None and still rely on `np.median`. The reason being that `np.median` is probably much faster than our more general `_weighted_percentile` (could be wrong on that, but I'd be surprised).

NicolasHug commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -500,18 +500,20 @@ def test_scorer_sample_weight():
-                                err_msg="scorer {0} behaves differently when "
-                                "ignoring samples and setting sample_weight to"
-                                " 0: {1} vs {2}".format(name, weighted,
+            if name != 'neg_median_absolute_error':
```
can you help me undestand why you'd need this?

Also if we need to ignore `neg_median_absolute_error`, it would be preferable to filter it out before the loop `for name, scorer in SCORERS.items():`

lucyleeow replied:
Since our data is binary classification data, with target being 1 or 0, there are only 2 possible values of `np.abs(y_pred - y_true)`, 1 or 0. With only 2 possible values it's likely that median after 'randomly' removing/weighting as zero 10 elements (out of 25) is the same as the median with all 25 elements.

To filter before the loop would you make a copy of `SCORER` and remove `neg_median_absolute_error` before the loop? Is there a better way?

lucyleeow replied:
I could instead use `make_regression` to create a `y` just for `neg_median_absolute_error`?

jnothman replied:
I think using regression targets here for regression metrics seems reasonable...?

TBH, I think this test is weird. Why are we testing the metric properties of scorers when really the unit that this file should be testing is the wrapper around the metric? We should only be testing here that `sample_weight` is correctly passed to the metric, regardless of the parameters to `make_scorer`.

lucyleeow replied:
Fair point. Though I would like to add more tests for `_weighted_percentile` wrt to weighted median. Specifically, that `_weighted_percentile(percentile=50)` with equal weights is same as `np.median` and also that the difference between sum of weights left and right of the weighted median are the smallest possible. Reference: https://en.wikipedia.org/wiki/Weighted_median#Properties
If you agree, should I add the tests to this PR or a different one?

Also is `test_gradient_boosting_loss_functions.py` the best place for these tests? Makes it a bit difficult to find.

With regard to this PR, should I use regression targets just for `neg_median_absolute_error` or all regression metrics?

jnothman replied:
> With regard to this PR, should I use regression targets just for neg_median_absolute_error or all regression metrics?

if it works for all, that would be more elegant.

> Also is test_gradient_boosting_loss_functions.py the best place for these tests? Makes it a bit difficult to find.

Not `sklearn/metrics/tests/test_regreession.py`?

lucyleeow replied:
Whoops I meant the tests for `_weighted_percentile`, which are in `test_gradient_boosting_loss_functions.py`. I would like to add to test for `_weighted_percentile`

lucyleeow replied:
starting here: https://github.com/scikit-learn/scikit-learn/blob/2d12bb57506d72a110eaedcf65783fd49a045854/sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py#L106

NicolasHug reviewed (commented):
Thanks @lucyleeow , made a quick (and incomplete) pass.

I'm wondering whether it's worth extending `_weighted_percentile`, or if it'd be easier and faster to just have our own version of `weighted_median`? I would assume that computing a weighted median is simpler and faster than having a general algorithm that works for every percentile, though I haven't looked into the details

lucyleeow committed:
suggestions

lucyleeow commented:
> 'm wondering whether it's worth extending `_weighted_percentile`, or if it'd be easier and faster to just have our own version of `weighted_median`?

Hmm potentially we don't have to take the cumsum of the entire weight array and just stop if the value >50. Then just use `arr[index we stopped at]`? (note current implementation takes the lower weighted median as we use `np.searchsorted()` with default `side='left'`)

glemaitre commented:
> I'm not sure if thats possible for weighted median. To calculate we need to take the cumsum of the weights then find the index where the cumsum/total weights is 0.5, which are the same steps for calculating any percentile.

I am not really good a those but computing the median should come back to some efficient sorting and take the `n // 2` element. However, in the case of a weighted median, it seems that you will have to sum the weights anyhow and then computing any percentile, including the median, will be as costly.

@NicolasHug are you aware of an algorithm where it would be more efficient for only computing the weighted median instead of the generic one proposed here?

jnothman commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,44 @@
-    sorted_idx = np.argsort(array)
+    if sample_weight is None:
+        sample_weight = np.ones_like(array)
+    array = np.squeeze(array)
```
Doing this seems inconsistent with the reference to axis=0 in the docstring, as any axis could be removed.

lucyleeow replied:
Yes I understand your concern. The problem is that for single output the scorer function feeds `y_true` and `y_pred` as shape (n_samples, 1) `median_absolute_error` e.g., in `test_scorer_sample_weight`:
```
                weighted = scorer(estimator[name], X_test, target,
                                  sample_weight=sample_weight)
```
the shape of `y_pred` and `y_true` fed to `median_absolute_error` are (25, 1). This makes it difficult to differential between 1D array and multioutput 2D array, where you want the median value of each 'column'.

The only case I can see having trouble is where you have multioutput but only 1 same (1, n_labels). I could change to `np.squeeze(array, axis=-1)` ?

lucyleeow replied:
That doesn't work. Did an if statement.

jnothman commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,44 @@
-    percentile_idx = np.clip(percentile_idx, 0, len(sorted_idx)-1)
-    return array[sorted_idx[percentile_idx]]
+    weight_cdf = stable_cumsum(sorted_weights, axis=0)
+    max_weight_cdf = np.take(weight_cdf, -1, axis=0)
```
```suggestion
    max_weight_cdf = weight_cdf[-1]
```

jnothman reviewed (dismissed):
Shouldn't there be interpolation where the percentile sought does not exactly match that present?

jnothman commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,44 @@
+        return array[()]
+    sorted_idx = np.argsort(array, axis=0)
+    sorted_weights = _take_along_axis(sample_weight, sorted_idx, axis=0)
+    sorted_array = _take_along_axis(array, sorted_idx, axis=0)
```
why is this needed rather than using the `sorted_idx` directly in looking up the `percentile_idx`?

lucyleeow replied:
I didn't know how to make this work for 2D array? Can I do a nested `take_along_axis`?

lucyleeow replied:
Fancy indexing!

jnothman reviewed (changes requested):
I don't mind the extension to `_weighted_percentile`, except that if it's for efficiency, I think storing `sorted_array` is more expensive than applying the existing `_weighted_percentile` columnwise in a Python loop.

lucyleeow commented:
> Shouldn't there be interpolation where the percentile sought does not exactly match that present?

Indeed the original PR implemented interpolated weighted median - explained very well here: https://github.com/scikit-learn/scikit-learn/pull/6217#issuecomment-174312229

Later on @lorentzenchr suggested to use `_weighted_percentile`: https://github.com/scikit-learn/scikit-learn/pull/6217#issuecomment-611930725

I guess this depends on which definition of median you wish to use, whether it 'should' be a value from the array. Happy to implement either and I actually already implemented the interpolated weighted median before I read down the PR comments.

lucyleeow committed:
dont store sorted array, suggestions

lucyleeow committed:
remove print

NicolasHug commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,46 @@
+    If ``array`` is 2D, compute weighted ``percentile`` along axis=0.
     """
-    sorted_idx = np.argsort(array)
+    if sample_weight is None:
+        sample_weight = np.ones_like(array)
```
Depending on whether this slows down the current version, we might want to have 2 versions: one when sample_weight is None, and one when it's not None.

Maybe wait until your code is mature to benchmark this though.

lucyleeow replied:
Oh I did what you suggested and used `np.median()` when `sample_weight=None`. I guess I can remove this. I think I kept it just in case/for robustness but since this is an internal function will remove.

lucyleeow committed:
remove none sampleweight

lucyleeow commented:
ping @jnothman

jnothman commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,43 @@
+    n_dim = array.ndim
+    if n_dim == 0:
+        return array[()]
+    # Remove trailing single-dimensional entries from the array shape
```
I suspect that if the returned percentile needs to be a scalar for column vector input, this should be the responsibility of the caller, not of weighted percentile. I wouldn't expect this kind of automatic squeezing on other aggregation along an axis in Scientific Python.

lucyleeow replied:
You're right. And I realise the problem was that `sample_weight` wasn't the same shape as `array`. fixed.

jnothman commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,43 @@
+        percentile = array[sorted_idx[percentile_idx]][0]
+    elif n_dim == 2:
+        col_index = np.arange(array.shape[1])
+        fancy_index_percentile = tuple([percentile_idx, col_index])
```
```suggestion
        fancy_index_percentile = (percentile_idx, col_index)
```
Or just use 
```py
        percentile_in_sorted = sorted_idx[percentile_idx, col_idx]
        percentile = array[percentile_in_sorted, col_idx]
```
for the next 4 lines

lucyleeow replied:
Much better, thanks. Unfamiliar with fancy indexing.

jnothman replied:
This is actually a Python syntax thing. `(a, b)` is literally shorthand for `tuple([a, b])`. and `x[a,b]` is literally shorthand for `x[(a,b)]`

lucyleeow replied:
Thanks!

jnothman commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,43 @@
-    return array[sorted_idx[percentile_idx]]
+    weight_cdf = stable_cumsum(sorted_weights, axis=0)
+    adjusted_percentile = (percentile / 100.) * weight_cdf[-1]
+    if n_dim == 1:
```
You can avoid some of these conditionals by just making the array 2d above (`array = np.reshape(array, (-1, 1))` and returning `percentile[0] if ndim == 1 else percentile` below.

lucyleeow committed:
fix logic

lucyleeow committed:
lint

lucyleeow committed:
lint

jnothman commented:
It would be good to move those tests to sklearn/utils/tests/. And yes, you
should add something there to check sample weight support (usual test
compares integer weights to replication of samples).

lucyleeow committed:
add tests, mv to utils

lucyleeow committed:
reg target for test

lucyleeow committed:
lint

lucyleeow committed:
odd size in test

lucyleeow committed:
wording

lucyleeow committed:
lint

jnothman commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -0,0 +1,66 @@
+    score = _weighted_percentile(y, sw, 50)
+    assert score == 1
+
+    # Check for multioutput
```
`_weighted_percentile` is not necessarily about "output" so this comment is a bit obscure.

I'd rather a test along the lines of:
```
def test_weighted_percentile_2d():
   X = ...
   sw = ...
   for each value of _weighted_percentile(X, sw):
       assert that _weighted_percentile(X[:, i], sw) matches it
```

jnothman reviewed (approved):
Otherwise LGTM!!

Please add an entry to the change log at `doc/whats_new/v0.24.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

lucyleeow committed:
fix test

lucyleeow commented:
So sorry @jnothman! I realised later that my change to `test_scorer_sample_weight` was incorrect. I needed to amend `_make_estimators` as well, otherwise the estimator was trained on the incorrect target (the classification y and not the regression y). Once I fixed this the other problems became apparent:

The `assert weighted != unweighted` is only guaranteed to work for `neg_median_absolute_error` if we remove/weight as 0 an odd number of elements. This is because it is possible that by chance you remove the same number of elements from left and right of the median such that the median remains the same. e.g.
```
diff = y_true - y_test
np.median(diff) == np.median(diff[10:])
```
This is what occurred, so I removed 11 elements instead of 10.

The `assert_almost_equal(weighted, ignored)` which checks that the score is the same with the weights and with the elements removed only works for `neg_median_absolute_error` if the number of elements is odd. As we take the 'lower' percentile (always value from the array, left of where the 0.50 weight value sits) whereas median takes the average of the middle elements if n is even. Thus I made sample size 100 instead of 101.

I understand this is a bit of gymnastics just to make the test work and demonstrates a little your point:

> TBH, I think this test is weird. Why are we testing the metric properties of scorers when really the unit that this file should be testing is the wrapper around the metric? We should only be testing here that sample_weight is correctly passed to the metric, regardless of the parameters to make_scorer.

lucyleeow committed:
better 2d test

lucyleeow committed:
Merge branch 'master' into median_abs_err

lucyleeow committed:
whats new

lucyleeow committed:
lint

lucyleeow commented:
ping @jnothman

lucyleeow committed:
merge master

glemaitre commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,36 @@
 import numpy as np
 
 from .extmath import stable_cumsum
+from sklearn.utils.fixes import _take_along_axis
```
It should be a relative import here

glemaitre replied:
```suggestion
from .fixes import _take_along_axis
```

glemaitre commented on sklearn/metrics/_regression.py:
```diff
@@ -335,7 +336,8 @@ def mean_squared_log_error(y_true, y_pred, *,
 @_deprecate_positional_args
-def median_absolute_error(y_true, y_pred, *, multioutput='uniform_average'):
+def median_absolute_error(y_true, y_pred, *, multioutput='uniform_average',
+                          sample_weight=None,):
```
Either remove the last comma or do

```python
def median_absolute_error(
    y_true, y_pred, *, multioutput='uniform_average',
    sample_weight=None,
):
```

lucyleeow replied:
comma removed

glemaitre commented on sklearn/metrics/_regression.py:
```diff
@@ -387,7 +394,11 @@ def median_absolute_error(y_true, y_pred, *, multioutput='uniform_average'):
+    if sample_weight is None:
+        output_errors = np.median(np.abs(y_pred - y_true), axis=0)
+    else:
+        output_errors = _weighted_percentile(np.abs(y_pred - y_true),
```
since `sample_weight` can be a list at this point, you will need to validate it.

```python
from ..utils.validation import _check_sample_weight
sample_weight = _check_sample_weight(sample_weight, y_pred)
```

glemaitre commented on sklearn/utils/fixes.py:
```diff
@@ -165,3 +165,38 @@ class loguniform(scipy.stats.reciprocal):
+def _take_along_axis(arr, indices, axis):
+    """Implements a simplified version of numpy.take_along_axis if numpy
+    version < 1.15"""
+    import numpy
```
Numpy is already imported as `np`

glemaitre commented on sklearn/utils/fixes.py:
```diff
@@ -165,3 +165,38 @@ class loguniform(scipy.stats.reciprocal):
+    version < 1.15"""
+    import numpy
+
+    if numpy.__version__ >= LooseVersion('1.15'):
```
`np_version` is already containing the version

```suggestion
    if np_version > (1, 14):
```

glemaitre commented on sklearn/utils/fixes.py:
```diff
@@ -165,3 +165,38 @@ class loguniform(scipy.stats.reciprocal):
+                "`indices` and `arr` must have the same number of dimensions")
+
+        shape_ones = (1,) * indices.ndim
+        dest_dims = list(range(axis)) + [None] + list(range(axis+1,
+                                                            indices.ndim))
```
```suggestion
        dest_dims = (
            list(range(axis)) +
            [None] +
            list(range(axis + 1, indices.ndim))
        )
```

glemaitre commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,36 @@
 def _weighted_percentile(array, sample_weight, percentile=50):
     """
     Compute the weighted ``percentile`` of ``array`` with ``sample_weight``.
+    If ``array`` is 2D, compute weighted ``percentile`` along axis=0.
```
I think that we will need a proper docstring

```
"""Compute a weighted percentile.

Note that if `array` is a 2-D array, the `percentile` is computed along the axis 0.

Parameters
...
"""
```

glemaitre replied:
we need to mention that percentile is between 0 and 100

glemaitre commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,36 @@
+    if array.ndim == 1:
+        array = array.reshape((-1, 1))
+    if array.shape != sample_weight.shape:
+        sample_weight = sample_weight.reshape(array.shape)
```
Usually we expect `array` to be 2-D and `sample_weight` to be 1-D. It is going to blow up here:

```python
In [3]: # X is a 2-D array of 100 samples by 3 features 
   ...: X = np.random.randn(100, 3) 
   ...: # with its corresponding weight per samples 
   ...: sample_weight = np.random.randn(100) 
   ...:  
   ...: _weighted_percentile(X, sample_weight)                                                                                                                                                    
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-0716f7c3f0e0> in <module>
      4 sample_weight = np.random.randn(100)
      5 
----> 6 _weighted_percentile(X, sample_weight)

~/Documents/packages/scikit-learn/sklearn/utils/stats.py in _weighted_percentile(array, sample_weight, percentile)
     16         array = array.reshape((-1, 1))
     17     if array.shape != sample_weight.shape:
---> 18         sample_weight = sample_weight.reshape(array.shape)
     19     sorted_idx = np.argsort(array, axis=0)
     20     sorted_weights = _take_along_axis(sample_weight, sorted_idx, axis=0)

ValueError: cannot reshape array of size 100 into shape (100,3)

In [4]: _weighted_percentile(X, np.repeat(sample_weight.reshape(-1, 1), 3))
Out[4]: array([ 0.44473983, -0.38490779,  0.43873199])
```

One solution is to repeat `sample_weight` if `_num_samples(array) == _num_samples(sample_weight)`

lucyleeow replied:
Thanks I didn't know this. Should I add a test for this case?

glemaitre commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,36 @@
-    percentile_idx = np.clip(percentile_idx, 0, len(sorted_idx)-1)
-    return array[sorted_idx[percentile_idx]]
+    weight_cdf = stable_cumsum(sorted_weights, axis=0)
+    adjusted_percentile = (percentile / 100.) * weight_cdf[-1]
```
```suggestion
    adjusted_percentile = percentile / 100 * weight_cdf[-1]
```

glemaitre commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,36 @@
-    return array[sorted_idx[percentile_idx]]
+    weight_cdf = stable_cumsum(sorted_weights, axis=0)
+    adjusted_percentile = (percentile / 100.) * weight_cdf[-1]
+    percentile_idx = [np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
+                      for i in range(weight_cdf.shape[1])]
+    percentile_idx = np.array(percentile_idx)
```
```suggestion
    percentile_idx = np.array([
        np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
        for i in range(weight_cdf.shape[1])
    ])```

glemaitre commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -0,0 +1,77 @@
+    median = np.median(x_manual)
+    w_median = _weighted_percentile(x, weights)
+
+    assert median == w_median
```
whenever we have some floating number we should not use `==`.
For instance in array, we use `assert_allclose` and never `assert_array_equal`.

For a single number we do

```suggestion
    assert median == pytest.approx(w_median)
```

glemaitre commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -0,0 +1,77 @@
+    w_median = _weighted_percentile(x_2d, w_2d)
+
+    for i, value in enumerate(w_median):
+        assert(value == _weighted_percentile(x_2d[:, i], w_2d[:, i]))
```
assert should not take parenthesis:

```suggestion
        p = _weighted_percentile(x_2d[:, i], w_2d[:, i])
        assert value == pytest.approx(p)
```

glemaitre replied:
or

```
p_axis_0 = [
    _weighted_percentile(x_2d[:, i], w_2d[:, i]) for i in range(len(w_median))
]
assert_allclose(w_median, p_axis_0)
```

glemaitre commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def test_weighted_percentile_2d():
+    # Check for when array is 2D
```
I think that we should test the case -> `array` 2-D and `sample_weight` 1-D

glemaitre commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -0,0 +1,77 @@
+    sw = np.ones(102, dtype=np.float64)
+    sw[-1] = 0.0
+    score = _weighted_percentile(y, sw, 50)
+    assert score == 1
```
if we have a floating point number, just use `pytest.approx(...)` here and after.

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -85,11 +85,15 @@ def _require_positive_y(y):
     # some of the regressions scorers require strictly positive input.
-    sensible_regr.fit(X_train, y_train + 1)
+    if y_reg_train is None:
+        sensible_regr.fit(X_train, y_train + 1)
```
`y_train + 1` is to get only positive value? It seems that `_require_positive_y` would be much better no?

lucyleeow replied:
I think so, not sure exactly what the purpose of sometimes using values >=0 and other times using values >=1. Will amend.

lucyleeow replied:
@glemaitre Amended but maybe ` + 1` is faster?

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -480,39 +484,45 @@ def test_scorer_sample_weight():
+    X, y = make_classification(n_samples=101, random_state=0)
     _, y_ml = make_multilabel_classification(n_samples=X.shape[0],
                                              random_state=0)
+    _, y_reg = make_regression(n_samples=X.shape[0], n_features=X.shape[1],
```
It seems that `test_scorer_sample_weight` was designed to test classification score.

I think it would be best to create 2 tests function one dedicated to classification and one dedicated to regression. It will be easier to follow what is going on.

glemaitre commented:
Yes it would be best

On Tue, 26 May 2020 at 11:08, Lucy Liu <notifications@github.com> wrote:

> *@lucyleeow* commented on this pull request.
> ------------------------------
>
> In sklearn/utils/stats.py
> <https://github.com/scikit-learn/scikit-learn/pull/17225#discussion_r430266569>
> :
>
> >      """
> -    sorted_idx = np.argsort(array)
> +    n_dim = array.ndim
> +    if n_dim == 0:
> +        return array[()]
> +    if array.ndim == 1:
> +        array = array.reshape((-1, 1))
> +    if array.shape != sample_weight.shape:
> +        sample_weight = sample_weight.reshape(array.shape)
>
> Thanks I didn't know this. Should I add a test for this case?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/17225#discussion_r430266569>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABY32P4AWYANONOPUCQGUSDRTOBKRANCNFSM4NA5AWNA>
> .
>


-- 
Guillaume Lemaitre
Scikit-learn @ Inria Foundation
https://glemaitre.github.io/

lucyleeow committed:
suggestions, better docstring

lucyleeow committed:
weight 1d, amend test

lucyleeow committed:
split reg and clas in test sample weight

lucyleeow committed:
lint

lucyleeow commented:
Thanks for the thorough review @glemaitre. I've added the test for 2D array and 1D sample weight and split `test_scorer_sample_weight` into reg and clf versions.

lucyleeow committed:
use require positive

lucyleeow committed:
word

lucyleeow commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,75 @@ def test_scorer_sample_weight():
-                                err_msg="scorer {0} behaves differently when "
-                                "ignoring samples and setting sample_weight to"
-                                " 0: {1} vs {2}".format(name, weighted,
+        if name not in REGRESSION_SCORERS:
```
Wasn't sure how to do this outside of the for loop without copying the `SCORER` dict (@glemaitre)

glemaitre replied:
I think that this is fine. Another way could have been

```python
if name in CLF_SCORERS:
```

lucyleeow replied:
Annoyingly `CLF_SCORERS` does not overlap with `CLUSTER_SCORERS` and `MULTILABEL_ONLY_SCORERS`. 

(though `REGRESSION_SCORERS` overlaps with `REQUIRE_POSITIVE_Y_SCORERS` so I could adjust the classification test)

glemaitre replied:
Ah ok so it is good.

I might have done the following to avoid a level of indent and put a nice comment

```python
if name in REGRESSION_SCORERS:
    # skip the regression scores since we evaluate the classification scores
    continue

lucyleeow commented on sklearn/utils/stats.py:
```diff
@@ -1,18 +1,61 @@
+    if array.ndim == 1:
+        array = array.reshape((-1, 1))
+    # When sample_weight 1D, repeat for each array.shape[1]
+    if (array.shape != sample_weight.shape and
+            array.shape[0] == sample_weight.shape[0]):
```
Not sure of formatting of this (@glemaitre )

glemaitre replied:
That's good

glemaitre commented on sklearn/metrics/_regression.py:
```diff
@@ -387,7 +395,12 @@ def median_absolute_error(y_true, y_pred, *, multioutput='uniform_average'):
     y_type, y_true, y_pred, multioutput = _check_reg_targets(
         y_true, y_pred, multioutput)
-    output_errors = np.median(np.abs(y_pred - y_true), axis=0)
+    sample_weight = _check_sample_weight(sample_weight, y_pred)
```
```suggestion
```

glemaitre commented on sklearn/metrics/_regression.py:
```diff
@@ -387,7 +395,12 @@ def median_absolute_error(y_true, y_pred, *, multioutput='uniform_average'):
+    if sample_weight is None:
+        output_errors = np.median(np.abs(y_pred - y_true), axis=0)
+    else:
+        output_errors = _weighted_percentile(np.abs(y_pred - y_true),
```
```suggestion
        sample_weight = _check_sample_weight(sample_weight, y_pred)
        output_errors = _weighted_percentile(np.abs(y_pred - y_true),
```

glemaitre replied:
You need to add the checking here otherwise we will create an array of 1s.
Then we will never compute the median with NumPy.

glemaitre replied:
(This something which is written in the docstring of the `_check_sample_weight` but it could be tricky to find out :)

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,75 @@ def test_scorer_sample_weight():
+    # Odd number of test samples req for neg_median_absolute_error
+    X, y = make_regression(n_samples=101, n_features=20, random_state=0)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+    y_train = _require_positive_y(y_train)
```
it would be best to do this on the y

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,75 @@ def test_scorer_sample_weight():
+    # errors
+
+    # Odd number of test samples req for neg_median_absolute_error
+    X, y = make_regression(n_samples=101, n_features=20, random_state=0)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+    y_train = _require_positive_y(y_train)
+    y_test = _require_positive_y(y_test)
```
```suggestion
    X, y = make_regression(n_samples=101, n_features=20, random_state=0)
    y = _require_positive_y(y)
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
```

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,75 @@ def test_scorer_sample_weight():
+                    "scorer {0} behaves identically when "
+                    "called with sample weights: {1} vs "
+                    "{2}".format(name, weighted, unweighted))
+                assert_almost_equal(weighted, ignored,
+                                    err_msg="scorer {0} behaves differently "
+                                    "when ignoring samples and setting "
+                                    "sample_weight to 0: "
+                                    "{1} vs {2}".format(name, weighted,
                                                         ignored))
```
```suggestion
                assert_allclose(
                    weighted, ignored,
                    err_msg=f"scorer {name} behaves differently when ignoring "
                    f"samples and setting sample_weight to 0: {weighted} vs {ignored}"
                )
```

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,75 @@ def test_scorer_sample_weight():
-                "with sample weights: {1}".format(name, str(e)))
+            except TypeError as e:
+                assert "sample_weight" in str(e), (
+                    "scorer {0} raises unhelpful exception when called "
+                    "with sample weights: {1}".format(name, str(e)))
```
```suggestion
                    f"scorer {name} raises unhelpful exception when called "
                    f"with sample weights: {str(e)}"
                )
```

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,75 @@ def test_scorer_sample_weight():
+                    "scorer {0} behaves identically when "
+                    "called with sample weights: {1} vs "
+                    "{2}".format(name, weighted, unweighted))
+                assert_almost_equal(weighted, ignored,
```
I would do the same formatting than below and use an f-string

glemaitre reviewed (commented):
apart of style, it looks good to me

lucyleeow committed:
suggestions

lucyleeow committed:
formatting

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,70 @@ def test_scorer_sample_weight():
+            assert_almost_equal(weighted, ignored,
+                                err_msg=f"scorer {name} behaves differently "
+                                "when ignoring samples and setting "
+                                "sample_weight to 0: {weighted} vs {ignored}")
```
```suggestion
                                f"sample_weight to 0: {weighted} vs {ignored}")
```

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,70 @@ def test_scorer_sample_weight():
+                f"sample weights: {weighted} vs {unweighted}")
+            assert_almost_equal(weighted, ignored,
+                                err_msg=f"scorer {name} behaves differently "
+                                "when ignoring samples and setting "
```
```suggestion
                                f"when ignoring samples and setting "
```

glemaitre commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -493,31 +494,70 @@ def test_scorer_sample_weight():
-                                " 0: {1} vs {2}".format(name, weighted,
-                                                        ignored))
+                                err_msg=f"scorer {name} behaves differently "
+                                "when ignoring samples and setting "
+                                "sample_weight to 0: {weighted} vs {ignored}")
```
```suggestion
                                f"when ignoring samples and setting "
                                f"sample_weight to 0: {weighted} vs {ignored}")
```

glemaitre commented:
The last changes. I think that we strictly do not check this part of the string so the tests did not fail but we should not print the right error message (without the variable values)

lucyleeow committed:
fix typos
